activation/atxdb.go

```diff
@@ -94,6 +96,50 @@ func (m *ActivationDb) CalcActiveSetFromView(a *types.ActivationTx) (uint32, err
 
 }
 
+const NIPST_LAYERTIME = 1000
+
+func (db *ActivationDb) ValidateAtx(atx *types.ActivationTx) error {
+	// validation rules: no other atx with same id and sequence number
+	// if s != 0 the prevAtx is valid and it's seq num is s -1
+	// positioning atx is valid
+	// validate nipst duration?
+	// fields 1-7 of the atx are the challenge of the poet
+	// layer index i^ satisfies i -i^ < (layers_passed during nipst creation) ANTON: maybe should be ==?
+	// the atx view contains d active Ids
+
+	eatx, _ := db.GetAtx(atx.Id())
+	if eatx != nil {
+		return fmt.Errorf("Found atx with same id")
+	}
+	prevAtxIds, err := db.GetNodeAtxIds(atx.NodeId)
+	if err == nil && len(prevAtxIds) > 0 {
+		prevAtx, err := db.GetAtx(prevAtxIds[len(prevAtxIds)])
+		if err == nil {
+			if prevAtx.Sequence == atx.Sequence {
```

Comment:
We're only checking the last ATX - does `db.GetNodeAtxIds` return a **sorted** list?

Reply:
We don't check for a lower sequence number (e.g. `1 -> 2 -> 3 -> 2` would pass validation).

Reply:
We don't validate the `PrevATXId` field at all.

Reply:
returns a list of received atx's sorted by the time they were received

Reply:
The way I think this should be implemented is by storing the `nextATX` on the prev ATX (doubly linked list).

Then, to validate, you go to the linked `prevATX` and validate that its `nextATX` is blank and that the sequence numbers between those two are correct (current is greater than previous).

Reply:
Of course, we'd also need to verify that if no `prevATX` was linked - then no other ATXs exist for that miner ID.